crates/polars-utils/src/mmap.rs

```diff
@@ -0,0 +1,300 @@
+        self.data.len()
+    }
+
+    #[inline(always)]
+    pub fn position(&self) -> usize {
+        self.position
+    }
+
+    #[inline(always)]
+    pub fn from_slice(data: &[u8]) -> Self {
+        Self::new(MemSlice::from_slice(data))
+    }
+
+    #[inline(always)]
+    pub fn from_vec(data: Vec<u8>) -> Self {
+        Self::new(MemSlice::from_vec(data))
+    }
+
+    #[inline(always)]
+    pub fn from_reader<R: io::Read>(mut reader: R) -> io::Result<Self> {
+        let mut vec = Vec::new();
+        reader.read_to_end(&mut vec)?;
+        Ok(Self::from_vec(vec))
+    }
+
+    #[inline(always)]
+    pub fn read_slice(&mut self, n: usize) -> MemSlice {
+        let start = self.position;
+        let end = usize::min(self.position + n, self.data.len());
+        self.position = end;
+        self.data.slice(start, end)
+    }
+}
+
+impl io::Read for MemReader {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let n = usize::min(buf.len(), self.remaining_len());
+        buf[..n].copy_from_slice(&self.data[self.position..self.position + n]);
+        self.position += n;
+        Ok(n)
+    }
+}
+
+impl io::Seek for MemReader {
+    fn seek(&mut self, pos: io::SeekFrom) -> io::Result<u64> {
+        let position = match pos {
+            io::SeekFrom::Start(position) => usize::min(position as usize, self.total_len()),
+            io::SeekFrom::End(offset) => {
+                let Some(position) = self.total_len().checked_add_signed(offset as isize) else {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Seek before to before buffer",
+                    ));
+                };
+
+                position
+            },
+            io::SeekFrom::Current(offset) => {
+                let Some(position) = self.position.checked_add_signed(offset as isize) else {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Seek before to before buffer",
+                    ));
+                };
+
+                position
+            },
+        };
+
+        self.position = position;
+
+        Ok(position as u64)
+    }
+}
+
+impl MemSlice {
+    #[inline(always)]
+    pub fn to_vec(self) -> Vec<u8> {
+        <[u8]>::to_vec(self.deref())
+    }
+
+    #[inline]
+    pub fn from_vec(v: Vec<u8>) -> Self {
+        let end = v.len();
+
+        Self(MemSliceInner::Allocated(AllocatedSlice {
+            data: v.into_boxed_slice().into(),
+            start: 0,
+            end,
+        }))
+    }
+
+    #[inline]
+    pub fn from_slice(slice: &[u8]) -> Self {
+        let end = slice.len();
+        Self(MemSliceInner::Allocated(AllocatedSlice {
+            data: slice.into(),
+            start: 0,
+            end,
+        }))
+    }
+
+    /// Attempt to prefetch the memory belonging to to this [`MemSlice`]
+    #[inline]
+    pub fn prefetch(&self) {
```

Comment:
I really doubt whether it can work as expected.

`_mm_prefetch` prefetches data from RAM to cache. If the data is not yet present in RAM (i.e., mmap), it's a no-op. For mmap, I think `madvise` should be used instead.

Reply:
I tried madvise, no-op reading and prefetching. Prefetching was the only one that was faster over the memcpy approach.

Reply:
I did not notice a difference between no-op and prefetch_l2. May I ask how you benched it?

Reply:
For this, I think cold performance one large files is the most important. So I am running a `drop_caches` before every run and running on a dataset I get from someone of 11GB.

```bash
hyperfine --warmup 2 \
    -p 'sync; echo 3 | sudo tee /proc/sys/vm/drop_caches' './plparbench-before' \
    -p 'sync; echo 3 | sudo tee /proc/sys/vm/drop_caches' './plparbench-after'  \
    --runs 5
```

I have done the same with while preparing with `pcu-fadvise` but it seemed to cause the same benchmarks.

Warm performance does not really matter here.

Reply:
I wonder whether the input parquet is compressed. I believe the mmap code path is only used when reading uncompressed parquet.

For mmap, I think `madvise` ([`Mmap::advise`](https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#method.advise)) should be used instead of `fadvise`.

On my machine, with a 9GB uncompressed parquet, there is no difference between no-op and `prefetch_l2`. `madvise` is actually slower. I think it is because not the whole file content is used when parsing parquet, and it is not worthy to eagerly read the file into memory.

Reply:
> I wonder whether the input parquet is compressed. I believe the mmap code path is only used when reading uncompressed parquet.

I don't think this is true. The MmapSlice are used for both uncompressed and compressed pages at the moment.

> For mmap, I think madvise ([Mmap::advise](https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#method.advise)) should be used instead of fadvise.

I meant `pcu-fadvise` as an alternative for `drop_caches`. As a file eviction tool.

> I think it is because not the whole file content is used when parsing parquet, so it is not worthy to do aggressive pre-reading or prefetching.

At the moment, we prefetch per `ColumnChunk`. All this data should be used unless you request a limited amount of rows.

But I agree with you. This seems like faulty benchmarking on my side. I will investigate further later. I also want to test [`Mmap::lock`](https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#method.lock). Important is that we have both mmap and heap-allocated support. The mmap, even without prefetching, seems to have some performance benefits over memcpying mmap-ed data to the heap.

Reply:
> I also want to test [Mmap::lock](https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#method.lock).

It prevents memory from being paged to the swap area, and the amount of memory that a process can lock is limited by RLIMIT_MEMLOCK. I don't think this effect is desired. If you want to immediately read all pages into RAM, you can use [`PopulateRead`](https://docs.rs/memmap2/latest/memmap2/enum.Advice.html#variant.PopulateRead); however in my bench it is slower than no-op.

Reply:
> If you want to immediately read all pages into RAM, you can use [`PopulateRead`](https://docs.rs/memmap2/latest/memmap2/enum.Advice.html#variant.PopulateRead); however in my bench it is slower than no-op.

Yeah, I tried it as well just now. It is quite a bit slower.